eventuals/os.h

```diff
@@ -62,6 +66,177 @@ inline void SetAffinity(std::thread& thread, const size_t cpu) {
+      << "Failed to get stack address via 'pthread_attr_getstack(...)'";
+
+  PCHECK(pthread_attr_destroy(&attr) == 0)
+      << "Failed to destroy thread attributes via "
+         "'pthread_attr_destroy(...)'";
+
+  return stack_address;
+}
+#endif
+
+#ifdef __MACH__
+inline Bytes GetStackSize() {
+  return Bytes(pthread_get_stacksize_np(pthread_self()));
+}
+
+// For macOS `pthread_get_stackaddr_np` can return either the base
+// or the end of the stack.
+inline void* GetStackAddress() {
+  return pthread_get_stackaddr_np(pthread_self());
+}
+#endif
+
+#if defined(__x86_64__)
+// Implementation for Windows might be added in future for this architecture.
+// For x86 stack grows downward (from highest address to lowest).
+// (check_line_length skip)
+// https://eli.thegreenplace.net/2011/02/04/where-the-top-of-the-stack-is-on-x86/
+// So the stack end will be the lowest address.
+#ifdef __linux__
+inline void*
+GetStackStart(
+    void* stack_address,
+    const Bytes& stack_size) {
+  return (char*) stack_address + stack_size.bytes();
+}
+
+inline void* GetStackEnd(
+    void* stack_address,
+    const Bytes& stack_size) {
+  return stack_address;
+}
+#endif
+
+#ifdef __MACH__
+inline void* GetStackStart(
+    void* stack_address,
+    const Bytes& stack_size) {
+  [[maybe_unused]] int local_var{};
+  if (&local_var > stack_address) {
+    return (char*) stack_address - stack_size.bytes();
+  } else {
+    return stack_address;
+  }
+}
+
+inline void* GetStackEnd(
+    void* stack_address,
+    const Bytes& stack_size) {
+  [[maybe_unused]] int local_var{};
+  if (&local_var > stack_address) {
+    return stack_address;
+  } else {
+    return (char*) stack_address - stack_size.bytes();
+  }
+}
+#endif
+
+#ifndef _WIN32
+inline Bytes Available(
+    void* stack_start,
+    void* stack_end,
+    const size_t* latest_local_variable) {
+  return Bytes(
+      ((char*) latest_local_variable)
+      - ((char*) stack_end) - sizeof(latest_local_variable));
+}
+#endif
+#elif defined(i386) || defined(__i386__) || defined(__i386) || defined(_M_IX86)
+// Implementation for checking sufficient stack space might be added in future.
+#elif defined(__powerpc) || defined(__powerpc__) || defined(__powerpc64__)
+|| defined(__POWERPC__) || defined(__ppc__) || defined(__PPC__)
+    || defined(_ARCH_PPC)
+// Implementation for checking sufficient stack space might be added in future.
+#elif defined(__PPC64__) || defined(__ppc64__) || defined(_ARCH_PPC64)
+// Implementation for checking sufficient stack space might be added in future.
+#elif defined(__sparc__) || defined(__sparc)
+// Implementation for checking sufficient stack space might be added in future.
+#elif defined(__arm__) || defined(__arm64__)
+// Implementation for checking sufficient stack space might be added in future.
+#endif
+
+////////////////////////////////////////////////////////////////////////
+
+#ifndef _WIN32
+inline void CheckSufficientStackSpace(const size_t composed_size) {
```

Comment:
```suggestion
#ifndef _WIN32
inline void CheckSufficientStackSpace(const size_t size) {
```
Since we might call this for other things beyond just when we're composing.